lib/client/keyagent.go

```diff
@@ -258,77 +258,99 @@ func (a *LocalKeyAgent) UserRefusedHosts() bool {
-		}
-		return nil
+	err := certChecker.CheckHostKey(addr, remote, key)
+	if err != nil {
+		a.log.Debugf("Host validation failed: %v.", err)
+		return trace.Wrap(err)
 	}
-	key = cert.SignatureKey
-	// we are given a certificate. see if it was signed by any of the known_host keys:
+	a.log.Debugf("Validated host %v.", addr)
+	return nil
+}
+
+// checkHostCertificate validates a host certificate. First checks the
+// ~/.tsh/known_hosts cache and if not found, prompts the user to accept
+// or reject.
+func (a *LocalKeyAgent) checkHostCertificate(key ssh.PublicKey, addr string) bool {
+	// Check the local cache (where all Teleport CAs are placed upon login) to
+	// see if any of them match.
 	keys, err := a.keyStore.GetKnownHostKeys("")
 	if err != nil {
-		a.log.Error(err)
-		return trace.Wrap(err)
+		a.log.Errorf("Unable to fetch certificate authorities: %v.", err)
+		return false
 	}
-	a.log.Debugf("Got %d known hosts", len(keys))
 	for i := range keys {
-		if sshutils.KeysEqual(cert.SignatureKey, keys[i]) {
-			a.log.Debugf("Verified host %s", host)
-			return nil
+		if sshutils.KeysEqual(key, keys[i]) {
+			return true
 		}
 	}
-	// final step: if we have not seen the host key/cert before, lets ask the user if
-	// he trusts it, and add to the known_hosts if he says "yes"
-	if err = hostPromptFunc(host, key); err != nil {
-		// he said "no"
-		a.noHosts[host] = true
+
+	// If this certificate was not seen before, prompt the user essentially
+	// treating it like a key.
+	err = a.checkHostKey(addr, nil, key)
+	if err != nil {
+		return false
+	}
+	return true
+}
+
+// checkHostCertificate validates a host key. First checks the
+// ~/.tsh/known_hosts cache and if not found, prompts the user to accept
+// or reject.
+func (a *LocalKeyAgent) checkHostKey(addr string, remote net.Addr, key ssh.PublicKey) error {
+	var err error
+
+	// Check if this exact host is in the local cache.
+	keys, _ := a.keyStore.GetKnownHostKeys(addr)
+	if len(keys) > 0 && sshutils.KeysEqual(key, keys[0]) {
+		a.log.Debugf("Verified host %s.", addr)
+		return nil
+	}
+
+	// If this key was not seen before, prompt the user with a fingerprint.
+	if a.hostPromptFunc != nil {
+		err = a.hostPromptFunc(addr, key)
+	} else {
+		err = a.defaultHostPromptFunc(addr, key)
+	}
+	if err != nil {
+		a.noHosts[addr] = true
 		return trace.Wrap(err)
 	}
-	// user said "yes"
-	err = a.keyStore.AddKnownHostKeys(host, []ssh.PublicKey{key})
+
+	// If the user trusts the key, store the key in the local known hosts
+	// cache ~/.tsh/known_hosts.
+	err = a.keyStore.AddKnownHostKeys(addr, []ssh.PublicKey{key})
 	if err != nil {
-		a.log.Warn(err)
+		a.log.Warnf("Failed to save the host key: %v.", err)
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+// defaultHostPromptFunc is the default host key/certificates prompt.
+func (a *LocalKeyAgent) defaultHostPromptFunc(host string, key ssh.PublicKey) error {
+	userAnswer := "no"
+	if !a.noHosts[host] {
```

Comment:
I don't understand the rationale behind `a.noHosts` + not sure if we want to keep some local persistent concurrent cache of those? Can you come up with alternative design?

Reply:
This was just a straight port of the originally code. Looking into it, it doesn't appear to do anything. I'll remove it.

Reply:
Actually looked at it again, it's used in tests. Updated the comment for `noHosts` to reflect what it does and where it's used (tests).